docs/articles/concepts/joins.md

```diff
@@ -1,3 +1,775 @@
+keyed_emails = op.map("key_emails", emails, lambda x: (str(x["user_id"]), x["email"]))
+
+joined = op.join("join", keyed_names, keyed_emails, running=True)
+```
+
+Now let's run the dataflow again an inspect the output.
+
+```python
+op.inspect("check_join", joined)
+
+bytewax.testing.run_main(flow)
+```
+
+Here's what we get. Let's visualize the progress and outputs of the
+[`join`](/apidocs/bytewax.operators/index#bytewax.operators.join)
+state table again but with `running=True`.
+
+```{testoutput}
+join_eg.check_join: ('123', ('Bee', None))
+join_eg.check_join: ('123', ('Bee', 'bee@bytewax.io'))
+join_eg.check_join: ('456', ('Hive', None))
+join_eg.check_join: ('456', ('Hive', 'hive@bytewax.io'))
+join_eg.check_join: ('123', ('Bee', 'queen@bytewax.io'))
+```
+
+First the table starts empty.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+
+Then we see the first name value. Since the table was updated in any
+way, it emits the current values for that key and keeps them in the
+table! If it doesn't know a value yet, it fills it in as `None`.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | `"Bee"` | |
+
+```
+('123', ({'user_id': 123, 'name': 'Bee'}, None))
+```
+
+Then we see the second name value. The same thing happens
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | `"Bee"` | |
+| 456 | `"Hive"` | |
+
+```
+('456', ({'user_id': 456, 'name': 'Hive'}, None))
+```
+
+Now we see the first email value. The same rules apply, but now since
+there are values for all the sides, we see them all in the output. The
+state for that key is not cleared!
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | `"Bee"` | `"bee@bytewax.io"` |
+| 456 | `"Hive"` | |
+
+```
+('123', ('Bee', 'bee@bytewax.io'))
+```
+
+Then we see the second email value.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | `"Bee"` | `"bee@bytewax.io"` |
+| 456 | `"Hive"` | `"hive@bytewax.io"` |
+
+```
+('456', ('Hive', 'hive@bytewax.io'))
+```
+
+Note that none of the state we have seen has been cleared. This now
+means when we see the updated Bee email, we'll see some output!
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | `"Bee"` | `"queen@bytewax.io"` |
+| 456 | `"Hive"` | `"hive@bytewax.io"` |
+
+```
+('123', ('Bee', 'queen@bytewax.io'))
+```
+
+So the running join is cool in that you can track updates to changes
+in values over time. _But this comes with a downside!_ Because we
+never throw away the state for a key, this state keeps growing in
+memory _forever_ if you keep adding keys. This might be the behavior
+you want, but realize that it does not come for free. A non-running
+streaming join is better if you know you'll only get one value for
+each side for each key, since it knows it can discard the state after
```

Comment:
"it knows it can discard" sounds just a bit too black-magical-ish, let's change it to something like "it is set up to discard"